From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Muthukumar <mk@verveflux.com>
Date: Sun, 8 Feb 2026 06:16:40 +0530
Subject: feat: add override hooks for 10 APIs (Batch 4)

Add window.customOverrides hooks for crypto.subtle.unwrapKey,
CSS.supports, CSSStyleDeclaration.getPropertyValue,
DynamicsCompressorNode (threshold/knee/ratio/attack/release),
and Element (getBoundingClientRect/getClientRects).

diff --git a/third_party/blink/renderer/core/css/css.idl b/third_party/blink/renderer/core/css/css.idl
index 02f8f48f914abe68930c85910c9e611a7acc615e..02984f3f76ccc2ca28d37101cdc67e1f567617c2 100644
--- a/third_party/blink/renderer/core/css/css.idl
+++ b/third_party/blink/renderer/core/css/css.idl
@@ -33,7 +33,7 @@
     Exposed=Window,
     ImplementedAs=DOMWindowCSS
 ] namespace CSS {
-    [CallWith=ExecutionContext] boolean supports(DOMString property, DOMString value);
-    [CallWith=ExecutionContext] boolean supports(DOMString conditionText);
+    [CallWith=ScriptState] boolean supports(DOMString property, DOMString value);
+    [CallWith=ScriptState] boolean supports(DOMString conditionText);
     DOMString escape(DOMString ident);
 };
diff --git a/third_party/blink/renderer/core/css/css_style_declaration.cc b/third_party/blink/renderer/core/css/css_style_declaration.cc
index 46a30ec67345d06a6bdf70355d58dcc85f5ca3d7..4bc8e8fc6a2c9a26ee9ee6eb2d4484b77b2ad6b5 100644
--- a/third_party/blink/renderer/core/css/css_style_declaration.cc
+++ b/third_party/blink/renderer/core/css/css_style_declaration.cc
@@ -33,6 +33,7 @@
 #include <algorithm>
 
 #include "third_party/blink/renderer/bindings/core/v8/native_value_traits_impl.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
 #include "third_party/blink/renderer/core/css/css_primitive_value.h"
 #include "third_party/blink/renderer/core/css/css_property_names.h"
 #include "third_party/blink/renderer/core/css/css_style_declaration.h"
@@ -43,6 +44,7 @@
 #include "third_party/blink/renderer/core/css/property_bitsets.h"
 #include "third_party/blink/renderer/core/execution_context/execution_context.h"
 #include "third_party/blink/renderer/core/page/scrolling/sync_scroll_attempt_heuristic.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
 #include "third_party/blink/renderer/platform/bindings/exception_state.h"
 #include "third_party/blink/renderer/platform/bindings/script_state.h"
 #include "third_party/blink/renderer/platform/wtf/std_lib_extras.h"
@@ -166,6 +168,39 @@ void CSSStyleDeclaration::setCSSFloat(const ExecutionContext* execution_context,
                       exception_state);
 }
 
+String CSSStyleDeclaration::getPropertyValueForBinding(
+    ScriptState* script_state,
+    const String& property_name) {
+  String value = getPropertyValue(property_name);
+  if (!script_state || !script_state->ContextIsValid())
+    return value;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn = GetCustomOverrideFunction(
+      script_state, "cssStyleDeclaration", "getPropertyValueOverride");
+  if (override_fn.IsEmpty())
+    return value;
+
+  v8::Local<v8::Context> context = script_state->GetContext();
+  v8::Local<v8::Value> argv[] = {V8String(isolate, value)};
+  v8::Local<v8::Value> override_result;
+  if (!V8ScriptRunner::CallFunction(
+           override_fn, ExecutionContext::From(script_state),
+           context->Global(), 1, argv, isolate)
+           .ToLocal(&override_result))
+    return value;
+
+  if (override_result->IsNullOrUndefined())
+    return value;
+  v8::Local<v8::String> override_string;
+  if (!override_result->ToString(context).ToLocal(&override_string))
+    return value;
+  return ToCoreString(isolate, override_string);
+}
+
 String CSSStyleDeclaration::AnonymousNamedGetter(const AtomicString& name) {
   // Search the style declaration.
   CSSPropertyID unresolved_property =
diff --git a/third_party/blink/renderer/core/css/css_style_declaration.h b/third_party/blink/renderer/core/css/css_style_declaration.h
index ab1265e441e91a2b43e38c2a5b8789f01db879b0..4bdcd3fbb5f844ad70991ad033659ca1ae71c0bf 100644
--- a/third_party/blink/renderer/core/css/css_style_declaration.h
+++ b/third_party/blink/renderer/core/css/css_style_declaration.h
@@ -37,6 +37,7 @@ class CSSStyleSheet;
 class CSSValue;
 class ExceptionState;
 class ExecutionContext;
+class ScriptState;
 enum class SecureContextMode;
 
 class CORE_EXPORT CSSStyleDeclaration : public ScriptWrappable,
@@ -64,6 +65,7 @@ class CORE_EXPORT CSSStyleDeclaration : public ScriptWrappable,
   virtual unsigned length() const = 0;
   virtual String item(unsigned index) const = 0;
   virtual String getPropertyValue(const String& property_name) = 0;
+  String getPropertyValueForBinding(ScriptState*, const String& property_name);
   virtual String getPropertyPriority(const String& property_name) = 0;
   virtual String GetPropertyShorthand(const String& property_name) = 0;
   virtual bool IsPropertyImplicit(const String& property_name) = 0;
diff --git a/third_party/blink/renderer/core/css/css_style_declaration.idl b/third_party/blink/renderer/core/css/css_style_declaration.idl
index 8f1cee005a3359dc3e483e60d6910ae88e766e04..5213f7bf6a33ec213a51b70104f0c7f97c2df130 100644
--- a/third_party/blink/renderer/core/css/css_style_declaration.idl
+++ b/third_party/blink/renderer/core/css/css_style_declaration.idl
@@ -26,7 +26,7 @@
     [CEReactions, RaisesException=Setter, SetterCallWith=ExecutionContext] attribute DOMString cssText;
     readonly attribute unsigned long length;
     [Affects=Nothing] getter DOMString item(unsigned long index);
-    DOMString getPropertyValue(DOMString property);
+    [CallWith=ScriptState, ImplementedAs=getPropertyValueForBinding] DOMString getPropertyValue(DOMString property);
     DOMString getPropertyPriority(DOMString property);
     [CEReactions, RaisesException, CallWith=ExecutionContext] void setProperty(DOMString property, [LegacyNullToEmptyString] DOMString value, optional [LegacyNullToEmptyString] DOMString priority = "");
     // void setPropertyValue(DOMString property, [LegacyNullToEmptyString] DOMString value);
diff --git a/third_party/blink/renderer/core/css/dom_window_css.cc b/third_party/blink/renderer/core/css/dom_window_css.cc
index 2dc7bb5c84f6255c73816eb863b2c6134ef02193..b21f5558cc213d0825c809c7940e1476b724be1e 100644
--- a/third_party/blink/renderer/core/css/dom_window_css.cc
+++ b/third_party/blink/renderer/core/css/dom_window_css.cc
@@ -29,34 +29,73 @@
 
 #include "third_party/blink/renderer/core/css/dom_window_css.h"
 
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
 #include "third_party/blink/renderer/core/css/css_markup.h"
 #include "third_party/blink/renderer/core/css/css_property_value_set.h"
 #include "third_party/blink/renderer/core/css/parser/css_parser.h"
 #include "third_party/blink/renderer/core/css/parser/css_property_parser.h"
 #include "third_party/blink/renderer/core/css/properties/css_property.h"
 #include "third_party/blink/renderer/core/execution_context/execution_context.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
+#include "third_party/blink/renderer/platform/bindings/script_state.h"
 #include "third_party/blink/renderer/platform/heap/garbage_collected.h"
 #include "third_party/blink/renderer/platform/wtf/text/string_builder.h"
 #include "third_party/blink/renderer/platform/wtf/text/wtf_string.h"
 
 namespace blink {
 
-bool DOMWindowCSS::supports(const ExecutionContext* execution_context,
+namespace {
+
+bool MaybeBoolOverride(ScriptState* script_state,
+                       const char* override_name,
+                       bool value) {
+  if (!script_state || !script_state->ContextIsValid())
+    return value;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn =
+      GetCustomOverrideFunction(script_state, "css", override_name);
+  if (override_fn.IsEmpty())
+    return value;
+
+  v8::Local<v8::Context> context = script_state->GetContext();
+  v8::Local<v8::Value> argv[] = {v8::Boolean::New(isolate, value)};
+  v8::Local<v8::Value> override_result;
+  if (!V8ScriptRunner::CallFunction(
+           override_fn, ExecutionContext::From(script_state),
+           context->Global(), 1, argv, isolate)
+           .ToLocal(&override_result))
+    return value;
+
+  if (override_result->IsBoolean())
+    return override_result.As<v8::Boolean>()->Value();
+  return value;
+}
+
+}  // namespace
+
+bool DOMWindowCSS::supports(ScriptState* script_state,
                             const String& property,
                             const String& value) {
+  const ExecutionContext* execution_context =
+      ExecutionContext::From(script_state);
   CSSPropertyID unresolved_property =
       UnresolvedCSSPropertyID(execution_context, property);
   if (unresolved_property == CSSPropertyID::kInvalid) {
-    return false;
+    return MaybeBoolOverride(script_state, "supportsOverride", false);
   }
   if (unresolved_property == CSSPropertyID::kVariable) {
     auto* dummy_style =
         MakeGarbageCollected<MutableCSSPropertyValueSet>(kHTMLStandardMode);
     bool is_animation_tainted = false;
-    return CSSParser::ParseValueForCustomProperty(
+    bool result = CSSParser::ParseValueForCustomProperty(
                dummy_style, AtomicString("--valid"), value, false,
                execution_context->GetSecureContextMode(), nullptr,
                is_animation_tainted) != MutableCSSPropertyValueSet::kParseError;
+    return MaybeBoolOverride(script_state, "supportsOverride", result);
   }
 
 #if DCHECK_IS_ON()
@@ -67,14 +106,19 @@ bool DOMWindowCSS::supports(const ExecutionContext* execution_context,
   // This will return false when !important is present
   auto* dummy_style =
       MakeGarbageCollected<MutableCSSPropertyValueSet>(kHTMLStandardMode);
-  return CSSParser::ParseValue(dummy_style, unresolved_property, value, false,
-                               execution_context) !=
-         MutableCSSPropertyValueSet::kParseError;
+  bool result = CSSParser::ParseValue(dummy_style, unresolved_property, value,
+                                      false, execution_context) !=
+                MutableCSSPropertyValueSet::kParseError;
+  return MaybeBoolOverride(script_state, "supportsOverride", result);
 }
 
-bool DOMWindowCSS::supports(const ExecutionContext* execution_context,
+bool DOMWindowCSS::supports(ScriptState* script_state,
                             const String& condition_text) {
-  return CSSParser::ParseSupportsCondition(condition_text, execution_context);
+  const ExecutionContext* execution_context =
+      ExecutionContext::From(script_state);
+  bool result =
+      CSSParser::ParseSupportsCondition(condition_text, execution_context);
+  return MaybeBoolOverride(script_state, "supportsOverride", result);
 }
 
 String DOMWindowCSS::escape(const String& ident) {
diff --git a/third_party/blink/renderer/core/css/dom_window_css.h b/third_party/blink/renderer/core/css/dom_window_css.h
index 570a706f8b6bf933ff2a6c5373370ebe4918dae5..7739accc71f5ddf4fadb6d9133c79a1a3cd54997 100644
--- a/third_party/blink/renderer/core/css/dom_window_css.h
+++ b/third_party/blink/renderer/core/css/dom_window_css.h
@@ -36,16 +36,16 @@
 
 namespace blink {
 
-class ExecutionContext;
+class ScriptState;
 
 class CORE_EXPORT DOMWindowCSS : public ScriptWrappable {
   DEFINE_WRAPPERTYPEINFO();
 
  public:
-  static bool supports(const ExecutionContext*,
+  static bool supports(ScriptState*,
                        const String& property,
                        const String& value);
-  static bool supports(const ExecutionContext*, const String& condition_text);
+  static bool supports(ScriptState*, const String& condition_text);
   static String escape(const String& ident);
 
  private:
diff --git a/third_party/blink/renderer/core/dom/element.cc b/third_party/blink/renderer/core/dom/element.cc
index 5ec94dee53c53005803404dfab13b2641ef6a4a8..ee83e8955f5b6dde7519552681cdd9b707dc0bd8 100644
--- a/third_party/blink/renderer/core/dom/element.cc
+++ b/third_party/blink/renderer/core/dom/element.cc
@@ -40,6 +40,8 @@
 #include "third_party/blink/public/web/web_autofill_state.h"
 #include "third_party/blink/renderer/bindings/core/v8/dictionary.h"
 #include "third_party/blink/renderer/bindings/core/v8/frozen_array.h"
+#include "third_party/blink/renderer/bindings/core/v8/to_v8_traits.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
 #include "third_party/blink/renderer/bindings/core/v8/script_promise_resolver.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_aria_notification_options.h"
 #include "third_party/blink/renderer/bindings/core/v8/v8_check_visibility_options.h"
@@ -243,8 +245,10 @@
 #include "third_party/blink/renderer/core/view_transition/view_transition_utils.h"
 #include "third_party/blink/renderer/core/xlink_names.h"
 #include "third_party/blink/renderer/core/xml_names.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
 #include "third_party/blink/renderer/platform/bindings/dom_data_store.h"
 #include "third_party/blink/renderer/platform/bindings/exception_state.h"
+#include "third_party/blink/renderer/platform/bindings/script_state.h"
 #include "third_party/blink/renderer/platform/bindings/v8_dom_activity_logger.h"
 #include "third_party/blink/renderer/platform/bindings/v8_dom_wrapper.h"
 #include "third_party/blink/renderer/platform/bindings/v8_per_context_data.h"
@@ -3229,7 +3233,7 @@ void Element::ClientQuads(Vector<gfx::QuadF>& quads) const {
   }
 }
 
-DOMRectList* Element::getClientRects() {
+DOMRectList* Element::getClientRects(ScriptState* script_state) {
   Vector<gfx::RectF> rects = GetClientRectsNoAdjustment();
   if (rects.empty()) {
     return MakeGarbageCollected<DOMRectList>();
@@ -3240,7 +3244,25 @@ DOMRectList* Element::getClientRects() {
     GetDocument().AdjustRectForScrollAndAbsoluteZoom(rect,
                                                      *element_layout_object);
   }
-  return MakeGarbageCollected<DOMRectList>(rects);
+  DOMRectList* result = MakeGarbageCollected<DOMRectList>(rects);
+  if (script_state && script_state->ContextIsValid()) {
+    v8::Isolate* isolate = script_state->GetIsolate();
+    v8::HandleScope handle_scope(isolate);
+    ScriptState::Scope scope(script_state);
+    v8::Local<v8::Function> override_fn = GetCustomOverrideFunction(
+        script_state, "element", "getClientRectsOverride");
+    if (!override_fn.IsEmpty()) {
+      v8::Local<v8::Value> result_value =
+          ToV8Traits<DOMRectList>::ToV8(script_state, result);
+      v8::Local<v8::Value> argv[] = {result_value};
+      v8::Local<v8::Value> override_result;
+      std::ignore = V8ScriptRunner::CallFunction(
+          override_fn, ExecutionContext::From(script_state),
+          script_state->GetContext()->Global(), 1, argv, isolate)
+          .ToLocal(&override_result);
+    }
+  }
+  return result;
 }
 
 Vector<gfx::RectF> Element::GetClientRectsNoAdjustment() {
@@ -3295,11 +3317,29 @@ DOMRect* Element::GetBoundingClientRect() {
   return DOMRect::FromRectF(GetBoundingClientRectNoLifecycleUpdate());
 }
 
-DOMRect* Element::GetBoundingClientRectForBinding() {
+DOMRect* Element::GetBoundingClientRectForBinding(ScriptState* script_state) {
   // TODO(crbug.com/1499981): This should be removed once synchronized scrolling
   // impact is understood.
   SyncScrollAttemptHeuristic::DidAccessScrollOffset();
-  return GetBoundingClientRect();
+  DOMRect* result = GetBoundingClientRect();
+  if (script_state && script_state->ContextIsValid()) {
+    v8::Isolate* isolate = script_state->GetIsolate();
+    v8::HandleScope handle_scope(isolate);
+    ScriptState::Scope scope(script_state);
+    v8::Local<v8::Function> override_fn = GetCustomOverrideFunction(
+        script_state, "element", "getBoundingClientRectOverride");
+    if (!override_fn.IsEmpty()) {
+      v8::Local<v8::Value> result_value =
+          ToV8Traits<DOMRect>::ToV8(script_state, result);
+      v8::Local<v8::Value> argv[] = {result_value};
+      v8::Local<v8::Value> override_result;
+      std::ignore = V8ScriptRunner::CallFunction(
+          override_fn, ExecutionContext::From(script_state),
+          script_state->GetContext()->Global(), 1, argv, isolate)
+          .ToLocal(&override_result);
+    }
+  }
+  return result;
 }
 
 const AtomicString& Element::computedRole() {
diff --git a/third_party/blink/renderer/core/dom/element.h b/third_party/blink/renderer/core/dom/element.h
index cd4e25e6323a6144a4dab7b4b4b1c201c668cf35..d1f2cae405164cac338c57b7077bb3d680b19307 100644
--- a/third_party/blink/renderer/core/dom/element.h
+++ b/third_party/blink/renderer/core/dom/element.h
@@ -589,7 +589,7 @@ class CORE_EXPORT Element : public ContainerNode, public Animatable {
   // TODO(crbug.com/41417572): This method should replace the above method.
   gfx::Rect VisibleBoundsRespectingClipsInLocalRoot() const;
 
-  DOMRectList* getClientRects();
+  DOMRectList* getClientRects(ScriptState*);
   // Returns a list of clients Rects in zoomed pixel units.
   Vector<gfx::RectF> GetClientRectsNoAdjustment();
 
@@ -598,7 +598,7 @@ class CORE_EXPORT Element : public ContainerNode, public Animatable {
   // Returns a rectangle in CSS pixel units.  i.e. ignoring zoom.
   gfx::RectF GetBoundingClientRectNoLifecycleUpdate() const;
   DOMRect* GetBoundingClientRect();
-  DOMRect* GetBoundingClientRectForBinding();
+  DOMRect* GetBoundingClientRectForBinding(ScriptState*);
 
   // Call the NoLifecycleUpdate variants if you are sure that the lifcycle is
   // already updated to at least pre-paint clean.
diff --git a/third_party/blink/renderer/core/dom/element.idl b/third_party/blink/renderer/core/dom/element.idl
index f56d3be531dd38462c7ad665cbeb355c8af753ff..b9f8ad7e466baac7ab82a45d0b45aff9b5d06a09 100644
--- a/third_party/blink/renderer/core/dom/element.idl
+++ b/third_party/blink/renderer/core/dom/element.idl
@@ -130,8 +130,8 @@ dictionary SetHTMLUnsafeOptions {
 
     // CSSOM View Module
     // https://drafts.csswg.org/cssom-view/#extension-to-the-element-interface
-    [Measure] DOMRectList getClientRects();
-    [Affects=Nothing, Measure, RuntimeCallStatsCounter=ElementGetBoundingClientRect, ImplementedAs=GetBoundingClientRectForBinding] DOMRect getBoundingClientRect();
+    [CallWith=ScriptState, Measure] DOMRectList getClientRects();
+    [CallWith=ScriptState, Affects=Nothing, Measure, RuntimeCallStatsCounter=ElementGetBoundingClientRect, ImplementedAs=GetBoundingClientRectForBinding] DOMRect getBoundingClientRect();
 
     // https://drafts.csswg.org/cssom-view/#dom-element-checkvisibility
     [MeasureAs=ElementCheckVisibility] boolean checkVisibility(optional CheckVisibilityOptions options = {});
diff --git a/third_party/blink/renderer/core/input/touch_action_test.cc b/third_party/blink/renderer/core/input/touch_action_test.cc
index 2357209414f7c8b8afade39faeab9191aa154969..ac8656b86177822ed55bc991fb25715b4fcc809e 100644
--- a/third_party/blink/renderer/core/input/touch_action_test.cc
+++ b/third_party/blink/renderer/core/input/touch_action_test.cc
@@ -283,7 +283,7 @@ void TouchActionTest::RunTestOnTree(ContainerNode* root, WebView* web_view) {
     // elements with multiple border boxes with other elements in between. Use
     // the first border box (which we can easily visualize in a browser for
     // debugging).
-    Persistent<DOMRectList> rects = element->getClientRects();
+    Persistent<DOMRectList> rects = element->getClientRects(nullptr);
     ASSERT_GE(rects->length(), 0u) << failure_context.ToString();
     if (!rects->length())
       continue;
diff --git a/third_party/blink/renderer/modules/crypto/subtle_crypto.cc b/third_party/blink/renderer/modules/crypto/subtle_crypto.cc
index fa54ccd5b7156d7e1d06366b823df40d51255c0b..81d3890e4ca15371795030158337894c97e6f383 100644
--- a/third_party/blink/renderer/modules/crypto/subtle_crypto.cc
+++ b/third_party/blink/renderer/modules/crypto/subtle_crypto.cc
@@ -711,6 +711,10 @@ ScriptPromise<CryptoKey> SubtleCrypto::unwrapKey(
     bool extractable,
     const Vector<String>& raw_key_usages,
     ExceptionState& exception_state) {
+  // Check for override
+  if (auto ovr = MaybeSubtleCryptoOverride<CryptoKey>(script_state, "unwrapKeyOverride"))
+    return *ovr;
+
   // Method described by:
   // https://w3c.github.io/webcrypto/Overview.html#SubtleCrypto-method-unwrapKey
 
diff --git a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.cc b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.cc
index 98e22d0c1692ee9825255a71f22d4135aef970ab..9df81dd0194008f2aa3f226123a422927ef94bab 100644
--- a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.cc
+++ b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.cc
@@ -25,14 +25,18 @@
 
 #include "third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.h"
 
+#include "third_party/blink/renderer/bindings/core/v8/to_v8_traits.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_script_runner.h"
 #include "third_party/blink/renderer/bindings/modules/v8/v8_dynamics_compressor_options.h"
 #include "third_party/blink/renderer/modules/webaudio/audio_graph_tracer.h"
 #include "third_party/blink/renderer/modules/webaudio/audio_node_input.h"
 #include "third_party/blink/renderer/modules/webaudio/audio_node_output.h"
 #include "third_party/blink/renderer/platform/audio/audio_utilities.h"
 #include "third_party/blink/renderer/platform/audio/dynamics_compressor.h"
+#include "third_party/blink/renderer/platform/bindings/custom_override_helpers.h"
 #include "third_party/blink/renderer/platform/bindings/exception_messages.h"
 #include "third_party/blink/renderer/platform/bindings/exception_state.h"
+#include "third_party/blink/renderer/platform/bindings/script_state.h"
 #include "third_party/blink/renderer/platform/instrumentation/tracing/trace_event.h"
 
 namespace blink {
@@ -59,6 +63,31 @@ constexpr double kDefaultReleaseValue = 0.250;
 constexpr float kMinReleaseValue = 0.0f;
 constexpr float kMaxReleaseValue = 1.0f;
 
+void MaybeCallAudioParamOverride(ScriptState* script_state,
+                                 const char* override_name,
+                                 AudioParam* param) {
+  if (!script_state || !script_state->ContextIsValid())
+    return;
+
+  v8::Isolate* isolate = script_state->GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  ScriptState::Scope scope(script_state);
+
+  v8::Local<v8::Function> override_fn = GetCustomOverrideFunction(
+      script_state, "dynamicsCompressorNode", override_name);
+  if (override_fn.IsEmpty())
+    return;
+
+  v8::Local<v8::Value> param_value =
+      ToV8Traits<AudioParam>::ToV8(script_state, param);
+  v8::Local<v8::Value> argv[] = {param_value};
+  v8::Local<v8::Value> result;
+  std::ignore = V8ScriptRunner::CallFunction(
+      override_fn, ExecutionContext::From(script_state),
+      script_state->GetContext()->Global(), 1, argv, isolate)
+      .ToLocal(&result);
+}
+
 }  // namespace
 
 DynamicsCompressorNode::DynamicsCompressorNode(BaseAudioContext& context)
@@ -134,11 +163,11 @@ DynamicsCompressorNode* DynamicsCompressorNode::Create(
 
   node->HandleChannelOptions(options, exception_state);
 
-  node->attack()->setValue(options->attack());
-  node->knee()->setValue(options->knee());
-  node->ratio()->setValue(options->ratio());
-  node->release()->setValue(options->release());
-  node->threshold()->setValue(options->threshold());
+  node->attack(nullptr)->setValue(options->attack());
+  node->knee(nullptr)->setValue(options->knee());
+  node->ratio(nullptr)->setValue(options->ratio());
+  node->release(nullptr)->setValue(options->release());
+  node->threshold(nullptr)->setValue(options->threshold());
 
   return node;
 }
@@ -157,15 +186,19 @@ DynamicsCompressorNode::GetDynamicsCompressorHandler() const {
   return static_cast<DynamicsCompressorHandler&>(Handler());
 }
 
-AudioParam* DynamicsCompressorNode::threshold() const {
+AudioParam* DynamicsCompressorNode::threshold(ScriptState* script_state) const {
+  MaybeCallAudioParamOverride(script_state, "thresholdOverride",
+                              threshold_.Get());
   return threshold_.Get();
 }
 
-AudioParam* DynamicsCompressorNode::knee() const {
+AudioParam* DynamicsCompressorNode::knee(ScriptState* script_state) const {
+  MaybeCallAudioParamOverride(script_state, "kneeOverride", knee_.Get());
   return knee_.Get();
 }
 
-AudioParam* DynamicsCompressorNode::ratio() const {
+AudioParam* DynamicsCompressorNode::ratio(ScriptState* script_state) const {
+  MaybeCallAudioParamOverride(script_state, "ratioOverride", ratio_.Get());
   return ratio_.Get();
 }
 
@@ -173,11 +206,13 @@ float DynamicsCompressorNode::reduction() const {
   return GetDynamicsCompressorHandler().ReductionValue();
 }
 
-AudioParam* DynamicsCompressorNode::attack() const {
+AudioParam* DynamicsCompressorNode::attack(ScriptState* script_state) const {
+  MaybeCallAudioParamOverride(script_state, "attackOverride", attack_.Get());
   return attack_.Get();
 }
 
-AudioParam* DynamicsCompressorNode::release() const {
+AudioParam* DynamicsCompressorNode::release(ScriptState* script_state) const {
+  MaybeCallAudioParamOverride(script_state, "releaseOverride", release_.Get());
   return release_.Get();
 }
 
diff --git a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.h b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.h
index 24221db70c69bd76ee0cabf2d366b4d12d21078e..7eea65b279b3ea0f8629445345f8a2066690cdcc 100644
--- a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.h
+++ b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.h
@@ -36,6 +36,7 @@ namespace blink {
 
 class BaseAudioContext;
 class DynamicsCompressorOptions;
+class ScriptState;
 
 class MODULES_EXPORT DynamicsCompressorNode final : public AudioNode {
   DEFINE_WRAPPERTYPEINFO();
@@ -50,12 +51,12 @@ class MODULES_EXPORT DynamicsCompressorNode final : public AudioNode {
 
   void Trace(Visitor*) const override;
 
-  AudioParam* threshold() const;
-  AudioParam* knee() const;
-  AudioParam* ratio() const;
+  AudioParam* threshold(ScriptState*) const;
+  AudioParam* knee(ScriptState*) const;
+  AudioParam* ratio(ScriptState*) const;
   float reduction() const;
-  AudioParam* attack() const;
-  AudioParam* release() const;
+  AudioParam* attack(ScriptState*) const;
+  AudioParam* release(ScriptState*) const;
 
   // InspectorHelperMixin
   void ReportDidCreate() final;
diff --git a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.idl b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.idl
index 1d16e033cfe713bef7b38f18f808fa11f9995038..9640a1260101855f1ce83c1c349fe7ee8803e3f1 100644
--- a/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.idl
+++ b/third_party/blink/renderer/modules/webaudio/dynamics_compressor_node.idl
@@ -28,10 +28,10 @@
     Exposed=Window
 ] interface DynamicsCompressorNode : AudioNode {
     [RaisesException, Measure] constructor(BaseAudioContext context, optional DynamicsCompressorOptions options = {});
-    readonly attribute AudioParam threshold; // in Decibels
-    readonly attribute AudioParam knee; // in Decibels
-    readonly attribute AudioParam ratio; // unit-less
+    [CallWith=ScriptState] readonly attribute AudioParam threshold; // in Decibels
+    [CallWith=ScriptState] readonly attribute AudioParam knee; // in Decibels
+    [CallWith=ScriptState] readonly attribute AudioParam ratio; // unit-less
     readonly attribute float reduction; // in Decibels
-    readonly attribute AudioParam attack; // in Seconds
-    readonly attribute AudioParam release; // in Seconds
+    [CallWith=ScriptState] readonly attribute AudioParam attack; // in Seconds
+    [CallWith=ScriptState] readonly attribute AudioParam release; // in Seconds
 };
